src/common/configurator.h

```diff
@@ -0,0 +1,158 @@
+
+namespace goby
+{
+namespace common
+{
+template <typename Config> class ConfiguratorInterface
+{
+  public:
+    void finalize()
+    {
+        if (!config_finalized_)
+        {
+            try
+            {
+                finalize_cfg();
+            }
+            catch (common::ConfigException& e)
+            {
+                handle_config_error(e);
+                throw;
+            }
+            config_finalized_ = true;
+        }
+
+        if (app3_configuration().debug_cfg())
+        {
+            std::cout << cfg().DebugString() << std::endl;
+            exit(EXIT_SUCCESS);
+        }
+    }
+
+    const Config& cfg() const
+    {
+        check_finalized();
+        return const_cfg();
+    }
+
+    const goby::protobuf::App3Config& app3_configuration() const
+    {
+        check_finalized();
+        return const_app3_configuration();
+    }
+
+  private:
+    void check_finalized() const
+    {
+        if (!config_finalized_)
+            throw(
+                common::ConfigException("Configuration is not finalized (call finalize() first)"));
+    }
+
+    virtual void finalize_cfg() {}
+    virtual const Config& const_cfg() const = 0;
+    virtual const goby::protobuf::App3Config& const_app3_configuration() const = 0;
+
+    virtual void handle_config_error(common::ConfigException& e) {}
+
+  private:
+    bool config_finalized_{false};
+};
+
+/// Implementation of ConfiguratorInterface for Google Protocol buffers
+template <typename Config> class ProtobufConfigurator : public ConfiguratorInterface<Config>
+{
+  public:
+    ProtobufConfigurator(int argc, char* argv[]);
+
+  protected:
+    // subclass can override to finalize the configuration at runtime
+    virtual void finalize_configuration(Config& cfg) {}
+
+  private:
+    void finalize_cfg() override
+    {
+        finalize_configuration(cfg_);
+        common::ConfigReader::check_required_cfg(cfg_);
+    }
+
+    const Config& const_cfg() const override { return cfg_; }
+    const goby::protobuf::App3Config& const_app3_configuration() const override
+    {
+        return cfg_.app();
+    }
+
+    void handle_config_error(common::ConfigException& e) override
+    {
+        // output all the available command line options
+        if (e.error())
+            std::cerr
+                << "Problem parsing configuration: use --help or --example_config for more help."
+                << std::endl;
+    }
+
+  private:
+    Config cfg_;
+    boost::program_options::options_description od_{"Allowed options"};
+};
+
+template <typename Config>
+ProtobufConfigurator<Config>::ProtobufConfigurator(int argc, char* argv[])
+{
+    //
+    // read the configuration
+    //
+    boost::program_options::variables_map var_map;
+    try
+    {
+        std::string application_name;
+
+        // we will check it later
+        bool check_required_cfg = false;
+        common::ConfigReader::read_cfg(argc, argv, &cfg_, &application_name, &od_, &var_map,
```

Comment:
Not super important but at the earliest time you can convert this `argc, argv` to a C++ list to pass it around I would.

Reply:
boost::program_options needs it. I think it's OK to pass around argc, argv in this particular case, rather than turn it in to std::vector<std::string> and then turn it back into an int and char*[].